File: milestone 2/python_modules/module02/ex0/ft_first_exception.cpp
#include <iostream>
#include <string>
#include <cstring>

int	check_temperature(std::string temp_str)
{
	int	typecasted;

	typecasted = 0;
	std::cout << "\n"
				<< "Testing temperature: " << temp_str << "\n";
	try
	{
		typecasted = std::stoi(temp_str);
		if (typecasted > 40)
		{
			throw std::out_of_range("hot");
		}
		else if (typecasted < 0)
		{
			throw std::out_of_range("cold");
		}
		else
		{
			std::cout << typecasted << "°C is perfect for plants!" << std::endl;
		}
	}
	catch (const std::invalid_argument &e)
	{
		std::cout << temp_str << " is not a valid number" << std::endl;
	}
	catch (const std::out_of_range &e)
	{
		if (std::strcmp(e.what(), "hot") == 0)
		{
			std::cout << typecasted << "°C is too hot for plants (max 40°C)" << std::endl;
		}
		else if (std::strcmp(e.what(), "cold") == 0)
		{
			std::cout << typecasted << "°C is too cold for plants (min 0°C)" << std::endl;
		}
	}
	return (typecasted);
}

int	main(void)
{
	std::cout << "=== Garden Temperature Checker ==="
				<< "\n";
	check_temperature("25");
	std::cout << "\n";
	check_temperature("abc");
	std::cout << "\n";
	check_temperature("100");
	std::cout << "\n";
	check_temperature("-50");
	std::cout << "\n\n"
				<< "All tests completed - program didn't crash!" << std::endl;
}

File: milestone 2/python_modules/module02/ex0/ft_first_exception.py
def check_temperature(temp_str) -> int:
    print(f"\nTesting temperature: {temp_str}")
    try:
        if not isinstance(temp_str, str):
            typecasted = int(temp_str)
        else:
            typecasted = int(temp_str)

        if typecasted > 40:
            raise ValueError("hot")
        elif typecasted < 0:
            raise ValueError("cold")
        else:
            print(f"{typecasted}°C is perfect for plants!")

    except ValueError as e:
        if str(e) == "hot":
            print(f"{typecasted}°C is too hot for plants (max 40°C)")
        elif str(e) == "cold":
            print(f"{typecasted}°C is too cold for plants (min 0°C)")
        else:
            print(f"{temp_str} is not a valid number")


if __name__ == "__main__":
    print("=== Garden Temperature Checker ===")
    print("")
    check_temperature("25")
    print("")
    check_temperature("abc")
    print("")
    check_temperature("100")
    print("")
    check_temperature("-50")
    print("")
    print("All tests completed - program didn't crash!")
